src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +72,32 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+static char *utoa_upper(unsigned value, char *str, int base)
+{
+	const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+	/* Check base is supported. */
+	if ((base < 2) || (base > 36)) {
+		str[0] = '\0';
+		return NULL;
+	}
+
+	/* Convert to string. Digits are in reverse order.  */
+	int l = 0;
```

Comment:
`size_t` - it cannot go negative, so let's not represent it with a type that can and induce UB in the process. Likewise, please give this a more descriptive name, not `l`.

Reply:
Renamed to `size_t offset` per next item. This index is not local to the loop, it is used as the upper bound in the next loop (reverser).

Reply:
It might be better named `length` then as it bounds the length of that secondary iteration and defines how many characters are valid in the intermediary buffers?

Reply:
Made `offset` loop-local (for in-register optimization) and introduced `total` instead. Could be `length`.

Reply:
`total` or `length` works - it's perhaps a little odd calling it `total` but "total length" makes sense and all that.